build.assets/tooling/cmd/changelog/main.go

```diff
@@ -0,0 +1,222 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/alecthomas/kingpin/v2"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/build.assets/tooling/lib/git"
+)
+
+var (
+	baseBranch = kingpin.Flag(
+		"base-branch",
+		"The base release branch to generate the changelog for.  It will be of the form branch/v*",
+	).Envar("BASE_BRANCH").String()
+
+	baseTag = kingpin.Flag(
+		"base-tag",
+		"The tag/version to generate the changelog from. It will be of the form vXX.Y.Z, e.g. v15.1.1",
+	).Envar("BASE_TAG").String()
+)
+
+// getForkedBranch will attempt to find a root branch for the current one that is in the format branch/v*
+func getForkedBranch(dir string) (string, error) {
+	forkPointRef, err := git.RunCmd(dir, "merge-base", "--fork-point", "HEAD")
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	fbranch, err := git.RunCmd(dir, "branch", "--list", "branch/v*", "--contains", forkPointRef, "--format", "%(refname:short)")
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	if fbranch == "" { // stdout is empty indicating the search failed
+		return "", trace.Errorf("could not find a valid root branch")
+	}
+	return fbranch, nil
+}
+
+// getBranch will return branch if parsed otherwise will attempt to find it
+// Branch should be in the format "branch/v*"
+func getBranch(branch, dir string) (string, error) {
+	if branch == "" { // flag and env var not set, attempt to find
+		// get ref
+		ref, err := git.RunCmd(dir, "symbolic-ref", "HEAD")
+		if err != nil {
+			return "", trace.Wrap(err, "not on a branch")
+		}
+
+		// remove prefix and ensure that branch is in expected format
+		branch, _ = strings.CutPrefix(ref, "refs/heads/")
+		if branch == ref {
+			return "", trace.Errorf("not on a branch: %s", ref)
```

Comment:
Nit: We usually do not use `trace.Errorf` directly, prefer one of the specific errors or default to "BadParameter".
```suggestion
			return "", trace.BadParameter("not on a branch: %s", ref)
```